samza-kv-inmemory/src/test/java/org/apache/samza/storage/kv/inmemory/TestInMemoryKeyValueStore.java

```diff
@@ -19,66 +19,501 @@
+    this.inMemoryKeyValueStore.put(key(0), null);
+
+    assertNull(this.inMemoryKeyValueStore.get(key(0)));
+    verify(this.putsCounter, times(2)).inc();
+    verify(this.deletesCounter).inc();
+    verify(this.bytesWrittenCounter).inc(key(0).length + value(0).length);
+  }
+
+  @Test
+  public void testPutAll() {
+    List<Entry<byte[], byte[]>> entries = new ArrayList<>();
+    for (int i = 0; i < 10; i++) {
+      entries.add(new Entry<>(key(i), value(i)));
+    }
+    this.inMemoryKeyValueStore.putAll(entries);
+
+    for (int i = 0; i < 10; i++) {
+      assertArrayEquals(value(i), this.inMemoryKeyValueStore.get(key(i)));
+    }
+    verify(this.putsCounter, times(10)).inc();
+    verify(this.bytesWrittenCounter, times(10)).inc(key(0).length + value(0).length);
+  }
+
+  @Test
+  public void testPutAllUpdate() {
+    // check that an existing value is overridden
+    this.inMemoryKeyValueStore.put(key(0), value(1234));
+    List<Entry<byte[], byte[]>> entries = new ArrayList<>();
+    for (int i = 0; i < 10; i++) {
+      entries.add(new Entry<>(key(i), value(i)));
+    }
+    this.inMemoryKeyValueStore.putAll(entries);
+
+    for (int i = 0; i < 10; i++) {
+      assertArrayEquals(value(i), this.inMemoryKeyValueStore.get(key(i)));
+    }
+    // 1 time for initial value to be overridden, 10 times for "regular" puts
+    verify(this.putsCounter, times(11)).inc();
+    verify(this.bytesWrittenCounter).inc(key(0).length + value(1234).length);
+    verify(this.bytesWrittenCounter, times(10)).inc(key(0).length + value(0).length);
+  }
+
+  @Test
+  public void testPutAllWithNull() {
+    List<Entry<byte[], byte[]>> entries = new ArrayList<>();
+    for (int i = 0; i < 10; i++) {
+      entries.add(new Entry<>(key(i), value(i)));
+    }
+    this.inMemoryKeyValueStore.putAll(entries);
+
+    List<Entry<byte[], byte[]>> deleteEntries = new ArrayList<>();
+    for (int i = 0; i < 3; i++) {
+      deleteEntries.add(new Entry<>(key(i), null));
+    }
+    this.inMemoryKeyValueStore.putAll(deleteEntries);
+
+    for (int i = 0; i < 10; i++) {
+      if (i < 3) {
+        assertNull(this.inMemoryKeyValueStore.get(key(i)));
+      } else {
+        assertArrayEquals(value(i), this.inMemoryKeyValueStore.get(key(i)));
+      }
+    }
+    // 10 times for "regular" puts, 3 times for deletion puts
+    verify(this.putsCounter, times(13)).inc();
+    // 10 "regular" puts all have same size for key/value
+    verify(this.bytesWrittenCounter, times(10)).inc(key(0).length + value(0).length);
+    verifyNoMoreInteractions(this.bytesWrittenCounter);
+    verify(this.deletesCounter, times(3)).inc();
+  }
+
+  @Test
+  public void testDelete() {
+    this.inMemoryKeyValueStore.put(key(0), value(0));
+    this.inMemoryKeyValueStore.delete(key(0));
+    assertNull(this.inMemoryKeyValueStore.get(key(0)));
+
+    /*
+     * There is a bug in which deletes are double counted in metrics. This deletesCounter should only be invoked once
```

Comment:
ty for noting the bug; will make it easy to fix when tests fail once the bug is fixed